File: NQueue.py
__author__ = 'fafu'
import time
class Solution:
    # @return a list of lists of string
    def solveNQueens(self, n):
        def __calc__(tmp,result,index,se):
            lens = len(tmp)
            for i in xrange(n):
                if i not in se and (len(tmp)==0 or abs(tmp[-1]-i)!=1):
                    tmp.append(i)
                    se.add(i)
                    if index+1==n:
                        result.append(tmp)
                    else:
                        __calc__(tmp,result,index+1,se)
                    se.discard(i)
                tmp = tmp[0:lens]
        if n<4:
            return []
        result = []
        __calc__([],result,0,set())
        newresult = []
        for i in result:
            res = [["." for m in xrange(n)] for a in xrange(n)]
            j = 0
            while j<n:
                res[j][i[j]]="Q"
                j+=1
            newresult.append(["".join(m) for m in res])
        return newresult
if __name__=="__main__":
    s = Solution()
    time.clock()
    s.solveNQueens(9)
    print time.clock()

File: NQueue.java
package check;

import java.util.*;

/**
 * User: fafu
 * Date: 14-12-30
 * Time: 13-19-${SECOND}
 * This class is
 */
public class NQueue {


    public List<List<String>> solveNQueens(int n) {
        if(n< 4){
            return null;
        }

        List<List<Integer>> result =new ArrayList<List<Integer>>();
        __calc__(new ArrayList<Integer>(),result, 0, new HashSet<Integer>(),n);
        List<List<String>>newresult =new ArrayList<List<String>>();
        for(List<Integer> i:result){
//        List<List<String> res =
//                Collections.fill();
//        j = 0
//        while j<n:
//        res[j][i[j]] = "Q"
//        j += 1
//        newresult.append(["".join(m) for m in res])
        }
        return newresult;
    }

    public void __calc__(List<Integer> tmp, List<List<Integer>> result, int index, Set se, int n) {
        int lens = tmp.size();
        for (int i = 0; i < n; i++) {
            if (se.contains(i) && (tmp.size() == 0 || Math.abs(tmp.get(tmp.size() - 1) - i) != 1)) {

                tmp.add(i);
                se.add(i);
                if (index + 1 == n)
                    result.add(tmp);
                else {
                    __calc__(tmp, result, index + 1, se, n);
                    se.remove(i);
                }
            }

        }
    }
}